### PR TITLE
feat: Implemented Auto-Update Electron UX

### LINKED ISSUE #2226: Implement Auto-Update UX for Composer Electron
Work with @DesignPolice to implement in-app UX for automatic application updates.

### PR DESCRIPTION
**Pull request by tonyanziano**
## Description

This PR implements the UX for auto-updates in Composer Electron.

## Task Item

closes #2226 

## Screenshots
**Update available**
update-available

**Update available (dialog minimized)**
update-available-minimized

**Update failed**
update-failed

**Update succeeded (user chose to install and restart)**
update-succeeded-1

**Update succeeded (user chose to download the update manually)**
update-succeeded-2

**Update unavailable**
update-unavailable

**Update in progress (total size not yet known)**
update-progress-1

**Update in progress**
update-progress-2

**New electron-server tests for app updater**
update-tests

**Client tests**
image

### CODE REVIEW

File: Composer/packages/client/setupTests.ts [hunk 1]
```diff
@@ -23,3 +23,6 @@ expect.extend({
     };
   },
 });
+
+// for tests using Electron IPC to talk to main process
+(window as any).ipcRenderer = { on: jest.fn() };
```
Comment (a-b-r-o-w-n) [inline Composer/packages/client/setupTests.ts:28]: is there a jest env setting that would add this?

File: Composer/packages/client/src/components/AppUpdater/index.tsx [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+    }
+  }, [downloadOption]);
+
+  const handleDownloadOptionChange = useCallback((_ev, option) => {
+    setDownloadOption(option);
+  }, []);
+
+  // listen for app updater events from main process
+  useEffect(() => {
+    ipcRenderer.on('app-update', (_event, name, payload) => {
```
Comment (a-b-r-o-w-n) [inline Composer/packages/client/src/components/AppUpdater/index.tsx:78]: Does this work in the browser?

File: Composer/packages/client/src/components/AppUpdater/styles.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+import { css } from '@emotion/core';
+import { IButtonStyles } from 'office-ui-fabric-react/lib/Button';
+import { IDialogContentStyles, IDialogFooterStyles } from 'office-ui-fabric-react/lib/Dialog';
+import { IModalStyles } from 'office-ui-fabric-react/lib/Modal';
+
+export const optionIcon = checked => css`
+  vertical-align: text-bottom;
+  font-size: 18px;
+  margin-right: 10px;
+  color: ${checked ? '#0078d4' : '#000'};
```
Comment (a-b-r-o-w-n) [inline Composer/packages/client/src/components/AppUpdater/styles.ts:13]: Can you use the fluent color variables here? I think `@ui-fabric/fluent-theme` exports them.

File: Composer/packages/client/src/components/AppUpdater/styles.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+export const dialogCopy = css`
+  margin: 0px;
+  color: #000;
+`;
+
+export const modal: Partial<IModalStyles> = {
+  main: {
+    maxWidth: '427px !important',
+    width: '427px !important',
```
Comment (a-b-r-o-w-n) [inline Composer/packages/client/src/components/AppUpdater/styles.ts:29]: Try using the `minWidth` and `maxWidth` prop on the dialog component.

File: Composer/packages/electron-server/package.json [hunk 5]
```diff
@@ -46,8 +45,10 @@
     "electron": "8.0.2",
     "electron-builder": "^22.4.1",
     "fs-extra": "^9.0.0",
+    "jest": "24.0.0",
```
Comment (a-b-r-o-w-n) [inline Composer/packages/electron-server/package.json:48]: Shouldn't need this if you are using test-utils!

File: Composer/packages/electron-server/src/electronWindow.ts [hunk 6]
```diff
@@ -21,20 +21,24 @@ export default class ElectronWindow {
 
   private constructor() {
     // Create the browser window.
+    const { width, height } = screen.getPrimaryDisplay().workAreaSize;
     const browserWindowOptions: Electron.BrowserWindowConstructorOptions = {
-      width: 800,
-      height: 600,
+      width: width * 0.75,
+      height: height * 0.9,
```
Comment (a-b-r-o-w-n) [inline Composer/packages/electron-server/src/electronWindow.ts:27]: 👨‍🔬

File: Composer/packages/electron-server/src/electronWindow.ts [hunk 7]
```diff
@@ -21,20 +21,24 @@ export default class ElectronWindow {
...
       show: false,
+      title: `Bot Framework Composer (v${app.getVersion()})`,
     };
     if (process.platform === 'linux' && !isDevelopment) {
       // workaround for broken .AppImage icons since electron-builder@21.0.1 removed .AppImage desktop integration
       // (https://github.com/electron-userland/electron-builder/releases/tag/v21.0.1)
       browserWindowOptions.icon = join(getUnpackedAsarPath(), 'resources/composerIcon_1024x1024.png');
     }
     this._currentBrowserWindow = new BrowserWindow(browserWindowOptions);
+    this._currentBrowserWindow.on('page-title-updated', ev => ev.preventDefault());
```
Comment (a-b-r-o-w-n) [inline Composer/packages/electron-server/src/electronWindow.ts:41]: What's this for?

Comment (tonyanziano) [reply Composer/packages/electron-server/src/electronWindow.ts:41]: In the initialization code for the main browser window, I explicitly set the title. Later on in the app lifecycle, there are times when Electron might try to update the browser title (this happens on the initial load of `index.html` and Electron will use whatever is in the `<head><title></title><head>` tag -- which doesn't include the version) and overwrite this explicitly set title.

This line prevents that behavior and leaves the explicitly set title.

File: Composer/packages/client/src/components/AppUpdater/index.tsx (hunk 2 above)
Comment (tonyanziano) [reply Composer/packages/client/src/components/AppUpdater/index.tsx:78]: Ah, good catch! I need to hide this component in the web app version. 😅

File: Composer/packages/client/setupTests.ts (hunk 1 above)
Comment (tonyanziano) [reply Composer/packages/client/setupTests.ts:28]: unfortunately not =/

File: Composer/packages/electron-server/src/electronWindow.ts (hunk 7 above)
Comment (a-b-r-o-w-n) [reply Composer/packages/electron-server/src/electronWindow.ts:41]: Ah ok. Would you mind just adding a brief comment stating that this prevents the overwriting of the window title?

Comment (tonyanziano) [reply Composer/packages/electron-server/src/electronWindow.ts]: done!

**Pull request merged**